src/relentless/optimize/objective.py

```diff
@@ -374,6 +389,284 @@ def compute(self, variables, directory=None):
+                            self, bonds, neighbors
+                        )
+
+                        neighbors.filter(bond_exclusion_filter)
+                    if (
+                        "1-3" in bonded_exclusions
+                        and snap.angles.N != 0
+                        and len(neighbors[:]) > 0
+                    ):
+                        angles = numpy.vstack(
+                            [snap.angles.group, numpy.flip(snap.angles.group, axis=1)],
+                        )
+
+                        angle_exclusion_filter = EnsembleAverage._cantor_pairing(
+                            self, angles[:, (0, -1)], neighbors
+                        )
+
+                        neighbors.filter(angle_exclusion_filter)
+                    if (
+                        "1-4" in bonded_exclusions
+                        and snap.dihedrals.N != 0
+                        and len(neighbors[:]) > 0
+                    ):
+                        dihedrals = numpy.vstack(
+                            [
+                                snap.dihedrals.group,
+                                numpy.flip(snap.dihedrals.group, axis=1),
+                            ],
+                        )
+
+                        dihedral_exclusion_filter = EnsembleAverage._cantor_pairing(
+                            self, dihedrals[:, (0, -1)], neighbors
+                        )
+
+                        neighbors.filter(dihedral_exclusion_filter)
+
+                filter_j_gt_i = neighbors[:, 1] > neighbors[:, 0]
+                neighbors.filter(filter_j_gt_i)
+                # pair contributions to the gradient
+                for i in snap.particles.types:
+                    for j in snap.particles.types:
+                        filter_ij = numpy.logical_and(
+                            type_masks[i][neighbors[:, 0]],
+                            type_masks[j][neighbors[:, 1]],
+                        )
+                        for var in variables:
+                            gradient[var] += numpy.sum(
+                                self.potentials.pair.derivative(
+                                    (i, j), var, x=neighbors.distances[filter_ij]
+                                )
+                            ) / len(traj)
+
+                # bond contributions to the gradient
+                if snap.bonds.N != 0:
+                    bond_type_map = {type: i for i, type in enumerate(snap.bonds.types)}
+                    for i in snap.bonds.types:
+                        bond_type_filter = bond_type_map[i] == snap.bonds.typeid
+
+                        bonds = snap.bonds.group[bond_type_filter]
+
+                        # Get positions for all pairs of bonded particles
+                        pos_1 = pos[bonds[:, 0]]
+                        pos_2 = pos[bonds[:, 1]]
+
+                        # Calculate distances
+                        dr = numpy.linalg.norm(pos_2 - pos_1, axis=1)
```

Comment:
I think you need to apply PBCs before you do this calculation. That would apply for all vectors in bonded interactions.

Reply:
We would need to unwrap these coordinates, right? Then there should be a requirement that we have the images in the trajectory?

Reply:
No, you shouldn't need to *unwrap* any coordinates for these calculations. All these interactions are computed under the minimum image convention, so you will need to *wrap* them.

I would wrap the particle positions when you load them so you know they lie in the freud box, then I would wrap the difference vectors to apply the minimum image convention.